Cryptic error when Docker base image doesn't exist

The command to get the image size (`docker inspect -f {{.Size}}`) fails if the image doesn't exist, hence the cryptic error.

```
Caused by: java.lang.NumberFormatException: For input string: ""
    at java.lang.NumberFormatException.forInputString (NumberFormatException.java:65)
    at java.lang.Long.parseLong (Long.java:702)
    at java.lang.Long.parseLong (Long.java:817)
    at com.google.cloud.tools.jib.docker.DockerClient.sizeOf (DockerClient.java:235)
    at com.google.cloud.tools.jib.builder.steps.SaveDockerStep.call (SaveDockerStep.java:57)
    at com.google.cloud.tools.jib.builder.steps.SaveDockerStep.call (SaveDockerStep.java:32)
    at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly (TrustedListenableFutureTask.java:125)
    at com.google.common.util.concurrent.InterruptibleTask.run (InterruptibleTask.java:69)
    at com.google.common.util.concurrent.TrustedListenableFutureTask.run (TrustedListenableFutureTask.java:78)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
    at java.lang.Thread.run (Thread.java:834)
```

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -136,7 +136,8 @@ public class DockerClient {
       imageTarball.writeTo(stdin);
 
     } catch (IOException ex) {
-      // Tries to read from stderr.
+      // Tries to read from stderr. Not using getStderrOutput(), as we want to show the error
+      // message from the tarball I/O write failure when reading from stderr fails.
       String error;
       try (InputStreamReader stderr =
           new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -27,6 +27,7 @@ import java.io.BufferedOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -140,11 +141,10 @@ public class DockerClient {
       try (InputStreamReader stderr =
           new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
         error = CharStreams.toString(stderr);
-
       } catch (IOException ignored) {
-        // This ignores exceptions from reading stderr and throws the original exception from
+        // This ignores exceptions from reading stderr and uses the original exception from
         // writing to stdin.
-        throw ex;
+        error = ex.getMessage();
       }
       throw new IOException("'docker load' command failed with error: " + error, ex);
     }
@@ -154,11 +154,8 @@ public class DockerClient {
       String output = CharStreams.toString(stdout);
 
       if (dockerProcess.waitFor() != 0) {
-        try (InputStreamReader stderr =
-            new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
-          throw new IOException(
-              "'docker load' command failed with output: " + CharStreams.toString(stderr));
-        }
+        throw new IOException(
+            "'docker load' command failed with error: " + getStderrOutput(dockerProcess));
       }
 
       return output;
@@ -180,20 +177,17 @@ public class DockerClient {
       ImageReference imageReference, Path outputPath, Consumer<Long> writtenByteCountListener)
       throws InterruptedException, IOException {
     Process dockerProcess = docker("save", imageReference.toString());
+
     try (InputStream stdout = new BufferedInputStream(dockerProcess.getInputStream());
-        NotifyingOutputStream fileStream =
-            new NotifyingOutputStream(
-                new BufferedOutputStream(Files.newOutputStream(outputPath)),
-                writtenByteCountListener)) {
-      ByteStreams.copy(stdout, fileStream);
+        OutputStream fileStream = new BufferedOutputStream(Files.newOutputStream(outputPath));
+        NotifyingOutputStream notifyingFileStream =
+            new NotifyingOutputStream(fileStream, writtenByteCountListener)) {
+      ByteStreams.copy(stdout, notifyingFileStream);
     }
 
     if (dockerProcess.waitFor() != 0) {
-      try (InputStreamReader stderr =
-          new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
-        throw new IOException(
-            "'docker save' command failed with output: " + CharStreams.toString(stderr));
-      }
+      throw new IOException(
+          "'docker save' command failed with error: " + getStderrOutput(dockerProcess));
     }
   }
 
@@ -205,7 +199,7 @@ public class DockerClient {
    * @param newImageReference the new image reference
    * @see <a
    *     href="https://docs.docker.com/engine/reference/commandline/tag/">https://docs.docker.com/engine/reference/commandline/tag/</a>
-   * @throws InterruptedException if the 'docker tag' process is interrupted.
+   * @throws InterruptedException if the 'docker tag' process is interrupted
    * @throws IOException if an I/O exception occurs or {@code docker tag} failed
    */
   public void tag(ImageReference originalImageReference, ImageReference newImageReference)
@@ -213,13 +207,9 @@ public class DockerClient {
     // Runs 'docker tag'.
     Process dockerProcess =
         docker("tag", originalImageReference.toString(), newImageReference.toString());
-
     if (dockerProcess.waitFor() != 0) {
-      try (InputStreamReader stderr =
-          new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
-        throw new IOException(
-            "'docker tag' command failed with error: " + CharStreams.toString(stderr));
-      }
+      throw new IOException(
+          "'docker tag' command failed with error: " + getStderrOutput(dockerProcess));
     }
   }
 
@@ -229,9 +219,16 @@ public class DockerClient {
    * @param imageReference the image to find the size of
    * @return the size in bytes
    * @throws IOException if an I/O exception occurs
+   * @throws InterruptedException if the 'docker inspect' process is interrupted
    */
-  public long sizeOf(ImageReference imageReference) throws IOException {
+  public long sizeOf(ImageReference imageReference) throws IOException, InterruptedException {
     Process sizeProcess = docker("inspect", "-f", "{{.Size}}", imageReference.toString());
+
+    if (sizeProcess.waitFor() != 0) {
+      throw new IOException(
+          "'docker inspect' command failed with error: " + getStderrOutput(sizeProcess));
+    }
+
     return Long.parseLong(
         CharStreams.toString(
                 new InputStreamReader(sizeProcess.getInputStream(), StandardCharsets.UTF_8))
@@ -241,5 +238,14 @@ public class DockerClient {
   /** Runs a {@code docker} command. */
   private Process docker(String... subCommand) throws IOException {
     return processBuilderFactory.apply(Arrays.asList(subCommand)).start();
+  }
+
+  private static String getStderrOutput(Process process) {
+    try (InputStreamReader stderr =
+        new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8)) {
+      return CharStreams.toString(stderr);
+    } catch (IOException ex) {
+      return "unknown (failed to read error message from stderr due to " + ex.getMessage() + ")";
+    }
   }
 }

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java
@@ -122,7 +122,6 @@ public class DockerClientTest {
   @Test
   public void testLoad_stdinFail_stderrFail() throws InterruptedException {
     DockerClient testDockerClient = new DockerClient(ignored -> mockProcessBuilder);
-    IOException expectedIOException = new IOException();
 
     Mockito.when(mockProcess.getOutputStream())
         .thenReturn(
@@ -130,7 +129,7 @@ public class DockerClientTest {
 
               @Override
               public void write(int b) throws IOException {
-                throw expectedIOException;
+                throw new IOException("I/O failed");
               }
             });
     Mockito.when(mockProcess.getErrorStream())
@@ -148,7 +147,7 @@ public class DockerClientTest {
       Assert.fail("Write should have failed");
 
     } catch (IOException ex) {
-      Assert.assertSame(expectedIOException, ex);
+      Assert.assertEquals("'docker load' command failed with error: I/O failed", ex.getMessage());
     }
   }
 
@@ -168,7 +167,7 @@ public class DockerClientTest {
       Assert.fail("Process should have failed");
 
     } catch (IOException ex) {
-      Assert.assertEquals("'docker load' command failed with output: error", ex.getMessage());
+      Assert.assertEquals("'docker load' command failed with error: error", ex.getMessage());
     }
   }
 
@@ -203,7 +202,7 @@ public class DockerClientTest {
       Assert.fail("docker save should have failed");
 
     } catch (IOException ex) {
-      Assert.assertEquals("'docker save' command failed with output: error", ex.getMessage());
+      Assert.assertEquals("'docker save' command failed with error: error", ex.getMessage());
     }
   }
 
@@ -264,6 +263,28 @@ public class DockerClientTest {
 
     } catch (IOException ex) {
       Assert.assertEquals("'docker tag' command failed with error: error", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testSize_fail() throws InterruptedException {
+    DockerClient testDockerClient =
+        new DockerClient(
+            subcommand -> {
+              Assert.assertEquals(Arrays.asList("inspect", "-f", "{{.Size}}", "image"), subcommand);
+              return mockProcessBuilder;
+            });
+    Mockito.when(mockProcess.waitFor()).thenReturn(1);
+
+    Mockito.when(mockProcess.getErrorStream())
+        .thenReturn(new ByteArrayInputStream("error".getBytes(StandardCharsets.UTF_8)));
+
+    try {
+      testDockerClient.sizeOf(ImageReference.of(null, "image", null));
+      Assert.fail("docker inspect should have failed");
+
+    } catch (IOException ex) {
+      Assert.assertEquals("'docker inspect' command failed with error: error", ex.getMessage());
     }
   }
 

